src/types/with_entities.rs

```diff
@@ -0,0 +1,171 @@
+use super::*;
+use std::ops::Range;
+
+pub struct MessageEntityRef<'a> {
+    message: &'a str,
+    range: Range<usize>,
+    entity: &'a MessageEntity,
+}
+
+fn convert_offsets(text: &str, ranges16: &[Range<usize>]) -> Vec<Range<usize>> {
+    let mut offset16 = 0;
+    let mut offset8 = 0;
+    let mut mapping = vec![usize::MAX; text.as_bytes().len() + 1];
```

Comment:
I still don't think this is a good implementation.

It requires allocating a buffer 8 times the size of the message. At most this should be ~32 KiB which IMO is quite a lot.

This also has problems when there are no entities or if all entities are at the start of the message -- in this impl you still need to iterate over all characters.